src/main/java/org/sagebionetworks/bridge/services/PermissionService.java

```diff
@@ -158,6 +172,120 @@ public void deletePermission(String appId, String guid) {
+                        removablePermissions.put(entityType, entityTypeMap);
+                    } else {
+                        Map<String, Set<AccessLevel>> entityTypeMap = removablePermissions.getOrDefault(entityType, new HashMap<>());
+                        entityTypeMap.put(entityId, oldPermissionEntityTypeMap.get(entityId));
+                        removablePermissions.put(entityType, entityTypeMap);
+                    }
+                }
+            } else {
+                removablePermissions.put(entityType, oldPermissions.get(entityType));
+            }
+        }
+        
+        List<Permission> currentPermissions = permissionDao.getPermissionsForUser(account.getAppId(), account.getId());
+        for (EntityType entityType : removablePermissions.keySet()) {
+            for (String entityId : removablePermissions.get(entityType).keySet()) {
+                for (AccessLevel accessLevel : removablePermissions.get(entityType).get(entityId)) {
+                    for (Permission permission : currentPermissions) {
+                        if (permission.getEntityType().equals(entityType) &&
+                                permission.getEntityId().equals(entityId) &&
+                                permission.getAccessLevel().equals(accessLevel)) {
+                            try {
+                                deletePermission(account.getAppId(), permission.getGuid());
+                            } catch (EntityNotFoundException e) {
+                                // While this shouldn't happen, if the permission does not exist then it's ok to skip.
+                            }
+                        }
+                    }
+                }
+            }
+        }
+        
+        for (EntityType entityType : additionalPermissions.keySet()) {
+            for (String entityId : additionalPermissions.get(entityType).keySet()) {
+                for (AccessLevel accessLevel : additionalPermissions.get(entityType).get(entityId)) {
+                    Permission newPermission = new Permission();
+                    newPermission.setAppId(account.getAppId());
+                    newPermission.setUserId(account.getId());
+                    newPermission.setEntityType(entityType);
+                    newPermission.setEntityId(entityId);
+                    newPermission.setAccessLevel(accessLevel);
+                    
+                    try {
+                        createPermission(account.getAppId(), newPermission);
+                    } catch (ConstraintViolationException e) {
+                        // It's possible that a duplicate permission already exists.
+                    }
+                }
+            }
+        }
+    }
+    
+    private Map<EntityType, Map<String, Set<AccessLevel>>> getPermissionsForRoles(Account account) {
+        String orgId = account.getOrgMembership();
+        if (orgId == null) {
+            return new HashMap<>();
+        }
```

Comment:
If orgId is null, do we still call getSponsoredStudyIds()? What happens if we pass in a null orgId to that call?

Do we need a unit test for this case? Are there any permissions that get generated or changed if the account is not in an org?

Reply:
If we pass a null orgId to getSponsoredStudyIds(), it will throw an exception. If an account is missing an orgId, then this will prevent getSponsoredStudyIds() from being called and just return an empty map of permissions. Since all the permissions granted through roles are in the context of the account's org, there would be no permissions anyways.

A new account can't be missing an orgId because it's enforced in the service. And it's not possible to update the orgId of an existing account. So the only case this covers is if an old account is missing an org and goes through a role update. I'm not sure that any accounts like this exist but it seems safer to include the check for it. There is a unit test to make sure no permissions are updated and the sponsor service isn't called now.